File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.cs
using System;

namespace SICP
{
		public static class Math
		{
				public static int Fact1(int n)
				{
						if(n == 1)
						{
						    return 1;
						}
						else
						{
								return n * Fact1(n-1);
						}
				}

				private static int Fact2Iter(int product, int counter, int maxCount)
				{
						if (counter > maxCount)
						{
								return product;
						}
						else
						{
								return Fact2Iter(counter*product, counter+1, maxCount);
						}
				}
				
				public static int Fact2(int n) => Fact2Iter(1, 1, n);

				public static void Main()
				{
						Console.WriteLine($"{Fact1(7)}");
						Console.WriteLine($"{Fact2(7)}");
				}
		}
}

// Start with
// csc factorial.cs && mono factorial.exe

File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.rs
fn fact1(n: i64) -> i64 {
    if n == 1 {
        return 1
    } else {
        return n * fact1(n-1)
    }
}

fn fact2(n: i64) -> i64 {
    return fact2iter(1, 1, n)
}

fn fact2iter(product: i64, counter: i64, max_count: i64) -> i64 {
    if counter > max_count {
        return product
    } else {
        return fact2iter(counter*product, counter+1, max_count)
    }
}

// fn main() {
//     print!("{}\n", fact2(7));
//     print!("{}\n", fact1(7));
// }
